File: sdk/ios/Sources/FrakSDK/Net/HTTPClient.swift
import Foundation

/// Declines every redirect: every URL we request is already ours, so a redirect means
/// misconfiguration, not something to follow.
private final class NoRedirectDelegate: NSObject, URLSessionTaskDelegate {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest
    ) async -> URLRequest? {
        nil
    }
}

/// A response body exceeded `HTTPClient.maxResponseBodyBytes`. `total` is the advertised
/// `Content-Length` when the header alone was enough to reject the response, or the actual byte
/// count when the cap was hit only after the body was read (Content-Length absent or lying).
struct ResponseTooLargeError: Error, Sendable {
    let total: Int64
}

/// The SDK's networking layer: one GET, over `URLSession`. Zero third-party deps.
struct HTTPClient: Sendable {
    /// A response that made it back, whatever its status.
    struct Response: Sendable {
        let status: Int
        let body: Data
        let retryAfterSeconds: Int?

        var isSuccess: Bool { (200..<300).contains(status) }

        /// Maps this response to the generic non-2xx error. Callers with a route-specific
        /// status meaning map that themselves and fall back to this for everything else.
        func toServerError() -> FrakError {
            .server(status: status, code: JSONDecoding.errorCode(in: body), retryAfterSeconds: retryAfterSeconds)
        }
    }

    static let maxRetryAfterSeconds = 300

    /// 1 MiB. Both responses this client ever reads — a merchant config resolve and a rewards
    /// list — are small JSON; generous headroom above any real payload while still bounding
    /// memory and what an oversized or misbehaving response could force into UserDefaults.
    static let maxResponseBodyBytes: Int64 = 1024 * 1024

    /// Wall-clock ceiling for a whole request, both attempts included. This is the SDK's
    /// single authoritative timeout mechanism: `Deadline.run` races the request against this
    /// one wall-clock bound and is what actually fires on a hang. The session's own timeouts
    /// (see `defaultSession`) are set well above this deadline, so they exist only as a
    /// defense-in-depth backstop against `URLSession` wedging, not as the mechanism that fires.
    /// Matches Android's effective per-request budget of ~20s, though the two platforms enforce
    /// it differently: Android's per-attempt timeouts are tight enough to be the mechanism that
    /// actually fires there, where here it is `Deadline.run`.
    static let overallDeadlineSeconds: TimeInterval = 20

    /// Set well above `overallDeadlineSeconds` so neither can ever be the mechanism that
    /// actually ends a request — `Deadline.run` always wins first. Not `.infinity`/unset: this
    /// still bounds the pathological case of `Deadline.run`'s own cancellation failing to
    /// unblock a wedged `URLSessionTask`.
    private static let sessionBackstopSeconds: TimeInterval = 60

    static let defaultSession: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = sessionBackstopSeconds
        configuration.timeoutIntervalForResource = sessionBackstopSeconds
        configuration.urlCache = nil
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        return URLSession(configuration: configuration)
    }()

    private let baseURL: String
    private let session: URLSession
    private let overallDeadlineSeconds: TimeInterval
    private let redirectDelegate = NoRedirectDelegate()
    // Self-contained here rather than threaded through from DefaultFrakClient: HTTPClient is
    // constructed before that init has a fully-built FrakLogger to hand it. Defaults to nil —
    // silent — until a caller passes one in.
    private let logger: FrakLogger?

    init(
        baseURL: String,
        session: URLSession = HTTPClient.defaultSession,
        overallDeadlineSeconds: TimeInterval = HTTPClient.overallDeadlineSeconds,
        logger: FrakLogger? = nil
    ) {
        self.baseURL = baseURL
        self.session = session
        self.overallDeadlineSeconds = overallDeadlineSeconds
        self.logger = logger
    }

    /// Issues a GET and returns the raw response, successful or not. `path` includes
    /// the leading slash; nil query values are dropped rather than sent empty.
    func get(_ path: String, query: [String: String?] = [:]) async throws -> Response {
        let request = try buildRequest(path: path, query: query)
        do {
            // Deadline wraps both attempts; giving the retry its own fresh window would
            // double the worst-case wait.
            return try await Deadline.run(seconds: overallDeadlineSeconds) {
                try await self.performWithRetry(request)
            }
        } catch is Deadline.Exceeded {
            throw FrakError.network(underlying: Deadline.Exceeded())
        }
    }

    /// Issues a POST and returns the raw response, successful or not.
    ///
    /// No transport-level retry, unlike `get`: a POST is not safe to replay blindly.
    /// Retrying is the caller's decision, and the tracking queue makes it with an
    /// idempotency key the backend can dedupe on.
    func post(_ path: String, body: Data, headers: [String: String] = [:]) async throws -> Response {
        var request = try buildRequest(path: path, query: [:])
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        for (name, value) in headers {
            request.setValue(value, forHTTPHeaderField: name)
        }
        let finalRequest = request
        do {
            return try await Deadline.run(seconds: overallDeadlineSeconds) {
                try await self.attempt(finalRequest)
            }
        } catch is Deadline.Exceeded {
            throw FrakError.network(underlying: Deadline.Exceeded())
        } catch let error as URLError where error.code == .cancelled {
            throw CancellationError()
        } catch is CancellationError {
            throw CancellationError()
        } catch let error as FrakError {
            throw error
        } catch {
            throw FrakError.network(underlying: error)
        }
    }

    /// A dropped/reset/timed-out connection is worth one retry; a trust or configuration failure
    /// will just fail identically again, so retrying only burns the deadline's remaining budget
    /// on a guaranteed repeat.
    private static func isTransient(_ error: URLError) -> Bool {
        switch error.code {
        case .networkConnectionLost, .notConnectedToInternet, .timedOut, .cannotConnectToHost,
            .cannotFindHost, .dnsLookupFailed:
            true
        default:
            false
        }
    }

    private func performWithRetry(_ request: URLRequest) async throws -> Response {
        do {
            return try await attempt(request)
        } catch let error as URLError where error.code == .cancelled {
            throw CancellationError()
        } catch let error as URLError where Self.isTransient(error) {
            // One retry: a pooled connection closed server-side while idle fails on next
            // use, indistinguishable from a real failure. Safe only because this is GET-only.
            // Short, jittered delay: retrying instantly, in lockstep, across every client
            // affected by the same blip recreates the load spike that caused it.
            try await Task.sleep(nanoseconds: Self.retryDelayNanoseconds())
            return try await retry(request)
        } catch let error as URLError {
            // Not in isTransient's allowlist — e.g. a trust or TLS configuration failure.
            throw FrakError.network(underlying: error)
        } catch is CancellationError {
            throw CancellationError()
        } catch let error as FrakError {
            // attempt() already throws a fully-formed FrakError (e.g. an oversized body); wrapping
            // it again here would bury the real cause inside a second .network(underlying:).
            throw error
        } catch {
            throw FrakError.network(underlying: error)
        }
    }

    private static func retryDelayNanoseconds() -> UInt64 {
        let baseMilliseconds: UInt64 = 100
        let jitterMilliseconds = UInt64.random(in: 0...200)
        return (baseMilliseconds + jitterMilliseconds) * 1_000_000
    }

    private func retry(_ request: URLRequest) async throws -> Response {
        do {
            return try await attempt(request)
        } catch let error as URLError where error.code == .cancelled {
            throw CancellationError()
        } catch is CancellationError {
            throw CancellationError()
        } catch let error as FrakError {
            throw error
        } catch {
            throw FrakError.network(underlying: error)
        }
    }

    private func attempt(_ request: URLRequest) async throws -> Response {
        let start = Date()
        do {
            let response = try await attemptUnlogged(request)
            logResult(request, status: response.status, start: start)
            return response
        } catch {
            logResult(request, status: nil, start: start)
            throw error
        }
    }

    /// DEBUG-level only, symmetric with Android's `HttpClient`. Logs method, host, path and
    /// status/duration — never the query string (a merchant id or the anonymous id can ride in
    /// it, e.g. `campaigns?anonymousId=…`) and never a header value (an auth token lives
    /// there). `os.Logger` interpolation stays `.private` even though a path alone is rarely
    /// sensitive, since a future route could add a path parameter without anyone revisiting
    /// this call site.
    private func logResult(_ request: URLRequest, status: Int?, start: Date) {
        guard let logger else { return }
        let durationMs = Int(Date().timeIntervalSince(start) * 1000)
        let method = request.httpMethod ?? "GET"
        let host = request.url?.host ?? "?"
        let path = request.url?.path ?? "?"
        let statusText = status.map(String.init) ?? "error"
        logger.debug("Frak \(method) \(host)\(path) -> \(statusText) (\(durationMs)ms)")
    }

    private func attemptUnlogged(_ request: URLRequest) async throws -> Response {
        // Not a true streaming read: `session.data(for:)` buffers the entire body before this
        // function runs, so unlike Android's `readBytesUpTo` this does not bound peak memory
        // during the read — a chunked response with no (or a lying) Content-Length can still be
        // buffered in full before the check below ever runs. What is guaranteed: an oversized
        // body is never returned to a caller that would persist it, and an honest large
        // response with a real Content-Length is rejected before the transfer finishes.
        let (data, response) = try await session.data(for: request, delegate: redirectDelegate)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw FrakError.network(underlying: URLError(.badServerResponse))
        }
        // `expectedContentLength` is -1 when the header is absent; not sufficient alone since a
        // chunked or lying response has no (or a false) Content-Length, so the post-buffer
        // check below is the real backstop against an oversized body.
        if httpResponse.expectedContentLength > Self.maxResponseBodyBytes {
            throw FrakError.network(underlying: ResponseTooLargeError(total: httpResponse.expectedContentLength))
        }
        if Int64(data.count) > Self.maxResponseBodyBytes {
            throw FrakError.network(underlying: ResponseTooLargeError(total: Int64(data.count)))
        }
        return Response(status: httpResponse.statusCode, body: data, retryAfterSeconds: retryAfterSeconds(httpResponse))
    }

    private func retryAfterSeconds(_ response: HTTPURLResponse) -> Int? {
        guard let header = response.value(forHTTPHeaderField: "Retry-After"),
            let seconds = Int(header.trimmingCharacters(in: .whitespaces))
        else {
            return nil
        }
        return min(max(seconds, 1), Self.maxRetryAfterSeconds)
    }

    /// Values are percent-encoded with `PercentEncoding` (RFC 3986); keys are compile-time
    /// constants and are not encoded.
    ///
    /// Not `URLComponents`: its `percentEncodedQuery` allows several RFC-3986-reserved
    /// characters (`!$&'()*+,;=`) through unescaped and passes `+` through literally rather
    /// than encoding it.
    private func buildRequest(path: String, query: [String: String?]) throws -> URLRequest {
        let present = query.compactMapValues { $0 }
        let urlString: String
        if present.isEmpty {
            urlString = baseURL + path
        } else {
            let encodedQuery = present.map { key, value in "\(key)=\(PercentEncoding.encode(value))" }
                .joined(separator: "&")
            urlString = "\(baseURL)\(path)?\(encodedQuery)"
        }
        guard let url = URL(string: urlString) else {
            throw FrakError.network(underlying: URLError(.badURL))
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(FrakSDKVersion.current, forHTTPHeaderField: FrakSDKVersion.headerName)
        return request
    }
}

File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/net/HttpClient.kt
// Opted in for `PercentEncoding`, which is `@InternalFrakApi`. Per file, not module-wide, so the
// marker still applies elsewhere.
@file:OptIn(InternalFrakApi::class)

package id.frak.sdk.net

import id.frak.sdk.FrakSdkVersion
import id.frak.sdk.InternalFrakApi
import id.frak.sdk.core.FrakError
import id.frak.sdk.core.FrakLogger
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.withTimeout
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import kotlin.random.Random

/**
 * A response body exceeded [HttpClient.MAX_RESPONSE_BODY_BYTES]. [total] is the bytes read so far,
 * or the advertised Content-Length when that alone was enough to reject the response.
 */
internal class ResponseTooLargeException(
    val total: Long,
) : IOException("Response body exceeded the maximum of ${HttpClient.MAX_RESPONSE_BODY_BYTES} bytes (was $total)")

/** GET/POST over [HttpURLConnection]; no OkHttp, zero runtime deps. */
internal class HttpClient(
    private val baseUrl: String,
    private val ioDispatcher: CoroutineDispatcher,
    private val open: (URL) -> HttpURLConnection = { it.openConnection() as HttpURLConnection },
    // Built as a DefaultFrakClient constructor default, before its init body runs; null is silent.
    private val logger: FrakLogger? = null,
    /** Fed the `Date` header of every response; null when nothing in this client's tree signs proofs. */
    private val serverClock: ServerClock? = null,
) {
    data class Response(
        val status: Int,
        val body: String,
        val retryAfterSeconds: Long?,
    ) {
        val isSuccess: Boolean get() = status in 200..299
    }

    /** Non-2xx statuses are returned, not thrown; null query values are dropped, not sent empty. */
    suspend fun get(
        path: String,
        query: Map<String, String?> = emptyMap(),
        headers: Map<String, String> = emptyMap(),
    ): Response {
        // Deadline wraps both attempts, else the retry would get its own fresh window.
        return withDeadline {
            // Inside the deadline block so a malformed origin surfaces as FrakError.Network.
            val url = urlOrThrow(buildUrl(path, query))
            try {
                attempt(url, headers, null)
            } catch (retryable: IOException) {
                if (!isTransient(retryable)) throw FrakError.Network(retryable)
                // One retry, safe only because this is GET-only: an idle pooled connection closed
                // server-side fails on next use indistinguishably from a real failure.
                delay(retryDelayMillis())
                try {
                    attempt(url, headers, null)
                } catch (failed: IOException) {
                    failed.addSuppressed(retryable)
                    throw FrakError.Network(failed)
                }
            }
        }
    }

    /** [java.net.UnknownHostException] is an [IOException], not a [java.net.SocketException] — hence its own arm. */
    private fun isTransient(error: IOException): Boolean =
        error is java.net.SocketException ||
            error is java.io.EOFException ||
            error is java.io.InterruptedIOException ||
            error is java.net.UnknownHostException

    /** Jittered: every client retrying in lockstep recreates the spike it reacted to. */
    private fun retryDelayMillis(): Long = RETRY_DELAY_BASE_MILLIS + Random.nextLong(RETRY_DELAY_JITTER_MILLIS)

    /** No retry, unlike [get]: the transport can't tell whether the request was read before dying. */
    suspend fun post(
        path: String,
        body: String,
        headers: Map<String, String> = emptyMap(),
    ): Response =
        withDeadline {
            val url = urlOrThrow(baseUrl + path)
            try {
                attempt(url, headers, body)
            } catch (failed: IOException) {
                throw FrakError.Network(failed)
            }
        }

    /** Only [TimeoutCancellationException] is mapped; a real `CancellationException` propagates untouched. */
    private suspend fun withDeadline(block: suspend () -> Response): Response =
        try {
            withTimeout(OVERALL_DEADLINE_MILLIS) { block() }
        } catch (expired: TimeoutCancellationException) {
            throw FrakError.Network(expired)
        }

    /** Runs as a child coroutine so cancellation can reach a blocked socket read via disconnect(). */
    private suspend fun attempt(
        url: URL,
        headers: Map<String, String>,
        body: String?,
    ): Response {
        val method = if (body == null) "GET" else "POST"
        val startMillis = System.currentTimeMillis()
        return try {
            val response = attemptUnlogged(url, headers, body)
            logResult(method, url, response.status, startMillis)
            response
        } catch (thrown: Exception) {
            // Exception, not Throwable: a JVM Error must propagate untouched
            logResult(method, url, status = null, startMillis)
            throw thrown
        }
    }

    /** DEBUG-level only. Never logs the query string or a header value — both can carry identifiers. */
    private fun logResult(
        method: String,
        url: URL,
        status: Int?,
        startMillis: Long,
    ) {
        if (logger == null) return
        val durationMillis = System.currentTimeMillis() - startMillis
        val statusText = status?.toString() ?: "error"
        logger.debug("Frak $method ${url.host}${url.path} -> $statusText (${durationMillis}ms)")
    }

    private suspend fun attemptUnlogged(
        url: URL,
        headers: Map<String, String>,
        body: String?,
    ): Response =
        coroutineScope {
            val connection = open(url)
            val work = async(ioDispatcher) { connection.perform(headers, body) }
            try {
                work.await()
            } catch (cancelled: CancellationException) {
                runCatching { connection.disconnect() }
                throw cancelled
            }
            // No disconnect() on success: draining and closing the stream returns it to the pool.
        }

    private fun HttpURLConnection.perform(
        headers: Map<String, String>,
        body: String?,
    ): Response {
        requestMethod = if (body == null) "GET" else "POST"
        connectTimeout = CONNECT_TIMEOUT_MILLIS
        readTimeout = READ_TIMEOUT_MILLIS
        // Every URL here is ours, so a redirect means misconfiguration, not something to follow.
        instanceFollowRedirects = false
        useCaches = false
        setRequestProperty("Accept", "application/json")
        setRequestProperty(FrakSdkVersion.HEADER_NAME, FrakSdkVersion.HEADER_VALUE)
        headers.forEach { (name, value) -> setRequestProperty(name, value) }

        if (body != null) {
            setRequestProperty("Content-Type", "application/json")
            doOutput = true
            val bytes = body.toByteArray(Charsets.UTF_8)
            setFixedLengthStreamingMode(bytes.size)
            outputStream.use { it.write(bytes) }
        }

        val status = responseCode
        // Free on every response, and the only clock the SDK can trust for proof timestamps.
        serverClock?.observe(getHeaderFieldDate("Date", 0L))
        // Fail fast on an advertised size; the read below caps chunked or lying responses too.
        if (contentLengthLong > MAX_RESPONSE_BODY_BYTES) {
            // Nothing was read, so disconnect() is the only way back to a clean pooled state.
            runCatching { errorStream?.close() }
            runCatching { inputStream?.close() }
            runCatching { disconnect() }
            throw FrakError.Network(ResponseTooLargeException(contentLengthLong))
        }
        // 204/205/304 carry no body, and inputStream throws for them rather than returning null
        if (status == 204 || status == 205 || status == 304) {
            runCatching { inputStream?.close() }
            runCatching { errorStream?.close() }
            return Response(status, "", retryAfterSeconds())
        }
        // Error bodies are on errorStream, not inputStream. Drained fully or the pooled connection is poisoned.
        val stream = if (status in 200..399) inputStream else errorStream
        val responseBody = stream?.use { readBytesUpTo(it, MAX_RESPONSE_BODY_BYTES) }.orEmpty()
        return Response(status, responseBody, retryAfterSeconds())
    }

    /** Bounded read, since Content-Length can be absent or wrong. Throws rather than truncating. */
    private fun HttpURLConnection.readBytesUpTo(
        stream: java.io.InputStream,
        limit: Long,
    ): String {
        val buffer = java.io.ByteArrayOutputStream()
        val chunk = ByteArray(8192)
        var total = 0L
        while (true) {
            val read = stream.read(chunk)
            if (read == -1) break
            total += read
            if (total > limit) {
                // Aborting mid-read leaves the stream undrained, so drop the connection.
                runCatching { disconnect() }
                throw FrakError.Network(ResponseTooLargeException(total))
            }
            buffer.write(chunk, 0, read)
        }
        return buffer.toString(Charsets.UTF_8.name())
    }

    /** Only the delta-seconds form is parsed; the limiter always emits one. */
    private fun HttpURLConnection.retryAfterSeconds(): Long? =
        getHeaderField("Retry-After")
            ?.trim()
            ?.toLongOrNull()
            ?.coerceIn(1L, MAX_RETRY_AFTER_SECONDS)

    private fun urlOrThrow(spec: String): URL =
        try {
            URL(spec)
        } catch (malformed: MalformedURLException) {
            throw FrakError.Network(malformed)
        }

    /** Builds the request URL. Values are percent-encoded; keys are compile-time constants and are not. */
    private fun buildUrl(
        path: String,
        query: Map<String, String?>,
    ): String {
        val present = query.mapNotNull { (key, value) -> value?.let { key to it } }
        if (present.isEmpty()) return baseUrl + path
        return present.joinToString(separator = "&", prefix = "$baseUrl$path?") { (key, value) ->
            "$key=${PercentEncoding.encode(value)}"
        }
    }

    companion object {
        /** Generic non-2xx mapping; callers with a route-specific status map that themselves first. */
        fun Response.toServerError(): FrakError.Server =
            FrakError.Server(status, JsonReader.errorCodeOrNull(body), retryAfterSeconds)

        /** Sized so two attempts plus [retryDelayMillis] fit inside [OVERALL_DEADLINE_MILLIS]. */
        const val CONNECT_TIMEOUT_MILLIS: Int = 3_000
        const val READ_TIMEOUT_MILLIS: Int = 5_000

        /** Wall-clock ceiling for a whole request; readTimeout alone only bounds each read. */
        const val OVERALL_DEADLINE_MILLIS: Long = 20_000

        const val RETRY_DELAY_BASE_MILLIS: Long = 100
        const val RETRY_DELAY_JITTER_MILLIS: Long = 200

        /** 5 minutes: long enough for any real rate limit, short enough to recover from a bad header. */
        const val MAX_RETRY_AFTER_SECONDS: Long = 300

        /** 1 MiB — generous headroom over the small JSON payloads this client reads. */
        const val MAX_RESPONSE_BODY_BYTES: Long = 1024L * 1024L
    }
}
